File: archives/C#/0725. Split Linked List in Parts.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode[] SplitListToParts(ListNode root, int k) {
        int rootLength=CalListLength(root);
        int k1=rootLength/k;//0
        int k2=rootLength%k;//3
        
        List<ListNode> result=new List<ListNode>();
        for(int i=0;i<k2;i++){
            ListNode[] rep=splitFirstKList(root,k1+1);
            root=rep[1];
            result.Add(rep[0]);
        }
        
       
        for(int i=0;i<k-k2;i++){
            ListNode[] rep1=splitFirstKList(root,k1);
            root=rep1[1];
            result.Add(rep1[0]);
        }
        return result.ToArray();
    }
    
    public int CalListLength(ListNode root){
        int cnt=0;
        while(root!=null){
            cnt++;
            root=root.next;
        }
        return cnt;
    }
    
    public ListNode[] splitFirstKList(ListNode root,int k){
        ListNode[] rep=new ListNode[]{null,null};
        if(root==null || k==0){
            return rep;
        }
        ListNode p=new ListNode(0);
        p.next=root;
        ListNode tmp=p;
        while(k>0){
            tmp=tmp.next;
            k--;
        }
        rep[1]=tmp.next;
        tmp.next=null;
        rep[0]=p.next;
        return rep;
    }
}

File: archives/20190519python/0725. Split Linked List in Parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def splitListToParts(self, root: ListNode, k: int) -> List[ListNode]:
        len_of_list=self.calLengthOfList(root)
        rep=[]
        if len_of_list<=k:
            while root:
                cur=root.next
                root.next=None
                rep.append(root)
                root=cur
                k-=1
            while k>0:
                rep.append(None)
                k-=1
        else:
            n=len_of_list//k
            fn=len_of_list%k
            self.getFirstNList(n,fn,rep,root,k)
        return rep
    
    def getFirstNList(self,n,fn,rep,root,k):
        if k==0:
            return 
        if fn>0:
            node=ListNode(0)
            node.next=root
            p=node
            cnt=n+1
            while cnt>0:
                cnt-=1
                p=p.next
            cur=p.next
            p.next=None
            rep.append(node.next)
            self.getFirstNList(n,fn-1,rep,cur,k-1)
        else:
            node=ListNode(0)
            node.next=root
            p=node
            cnt=n
            while cnt>0:
                cnt-=1
                p=p.next
            cur=p.next
            p.next=None
            rep.append(node.next)
            self.getFirstNList(n,fn-1,rep,cur,k-1)
    
    def calLengthOfList(self,root):
        k=0
        while root:
            k+=1
            root=root.next
        return k
